relay/nakama/auth/siwe.go

```diff
@@ -34,48 +34,87 @@ func validateSIWE(signer, message, signature string) (isValidationRequest bool,
 	return isValidationRequest, nil
 }
 
-func handleSIWE(
+func processSIWE(
 	ctx context.Context,
 	logger runtime.Logger,
 	nk runtime.NakamaModule,
-	in *api.AuthenticateCustomRequest,
-) (
-	*api.AuthenticateCustomRequest, error,
-) {
-	signerAddress := in.GetAccount().GetId()
-	signature := in.GetAccount().GetVars()["signature"]
-	message := in.GetAccount().GetVars()["message"]
-
+	signerAddress, signature, message string,
+) (bool, error) {
 	isValidationRequest, err := validateSIWE(signerAddress, message, signature)
 	if err != nil {
 		_, err = utils.LogErrorWithMessageAndCode(logger, err, codes.InvalidArgument, "invalid vars")
-		return nil, err
+		return false, err
 	}
 
 	if !isValidationRequest {
 		// The signature and message is empty. Generate a new SIWE message for the user.
 		resp, err := siwe.GenerateNewSIWEMessage(signerAddress)
 		if err != nil {
 			_, err = utils.LogError(logger, err, codes.FailedPrecondition)
-			return nil, err
+			return false, err
 		}
 
 		bz, err := json.Marshal(resp)
 		if err != nil {
 			_, err = utils.LogErrorWithMessageAndCode(logger, err, codes.FailedPrecondition, "")
-			return nil, err
+			return false, err
 		}
 		// An error is being returned here, but the error contains the text of the
 		// SIWE Message that needs to be signed.
-		return nil, runtime.NewError(string(bz), int(codes.Unauthenticated))
+		return false, runtime.NewError(string(bz), int(codes.Unauthenticated))
 	}
 
 	// The user has provided a signature and a message. Attempt to authenticate the user.
 	if err := siwe.ValidateSignature(ctx, nk, signerAddress, message, signature); err != nil {
 		_, err = utils.LogErrorWithMessageAndCode(
 			logger, err, codes.Unauthenticated, "authentication failed")
-		return nil, err
+		return false, err
 	}
+
 	// The user has successfully been authenticated
-	return in, nil
+	return true, nil
+}
+
+func handleSIWEforAuth(
+	ctx context.Context,
+	logger runtime.Logger,
+	nk runtime.NakamaModule,
+	in *api.AuthenticateCustomRequest,
+) (*api.AuthenticateCustomRequest, error) {
+	signerAddress := in.GetAccount().GetId()
+	signature := in.GetAccount().GetVars()["signature"]
+	message := in.GetAccount().GetVars()["message"]
+
+	success, err := processSIWE(ctx, logger, nk, signerAddress, signature, message)
+	if err != nil {
+		return nil, err
+	}
+
+	if success {
+		return in, nil
+	}
+
+	return nil, err
+}
+
+func handleSIWEforLink(
+	ctx context.Context,
+	logger runtime.Logger,
+	nk runtime.NakamaModule,
+	in *api.AccountCustom,
+) (*api.AccountCustom, error) {
+	signerAddress := in.GetId()
+	signature := in.GetVars()["signature"]
+	message := in.GetVars()["message"]
+
+	success, err := processSIWE(ctx, logger, nk, signerAddress, signature, message)
+	if err != nil {
+		return nil, err
+	}
+
+	if success {
+		return in, nil
+	}
+
+	return nil, err
```

Comment:
For readability, it's typically encouraged to have the last return statement be the "success" case.

Reply:
updated